src/main/scala/Ring.scala

```diff
@@ -0,0 +1,207 @@
+package testchipip
+
+import chisel3._
+import chisel3.util._
+
+import freechips.rocketchip.config.Parameters
+import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.subsystem.{SystemBus, SystemBusParams}
+import freechips.rocketchip.tilelink._
+import freechips.rocketchip.util.{HellaPeekingArbiter, BooleanToAugmentedBoolean}
+
+abstract class RingInternalIO[T <: Data](bundleType: NetworkBundle[T]) extends Bundle {
+  val int_in = Flipped(Decoupled(bundleType))
+  val int_out = Decoupled(bundleType)
+}
+
+class RingInputNodeIO[T <: Data](bundleType: NetworkBundle[T])
+    extends RingInternalIO(bundleType) {
+  val ext_in = Flipped(Decoupled(bundleType))
+
+  override def cloneType =
+    new RingInputNodeIO(bundleType).asInstanceOf[this.type]
+}
+
+class RingOutputNodeIO[T <: Data](bundleType: NetworkBundle[T])
+    extends RingInternalIO(bundleType) {
+  val ext_out = Decoupled(bundleType)
+
+  override def cloneType =
+    new RingOutputNodeIO(bundleType).asInstanceOf[this.type]
+}
+
+class RingInputNode[T <: Data](
+    nNodes: Int, payloadTyp: T, buffer: BufferParams) extends Module {
+  val bundleType = new NetworkBundle(nNodes, payloadTyp)
+  val io = IO(new RingInputNodeIO(bundleType))
+
+  val extArb = Module(new HellaPeekingArbiter(
+    bundleType, 2, (b: NetworkBundle[T]) => b.last))
+  extArb.io.in <> Seq(io.ext_in, io.int_in)
+  io.int_out <> buffer(extArb.io.out)
+}
+
+class RingOutputNode[T <: Data](
+    id: Int, nNodes: Int, payloadTyp: T, buffer: BufferParams) extends Module {
+  val bundleType = new NetworkBundle(nNodes, payloadTyp)
+  val io = IO(new RingOutputNodeIO(bundleType))
+
+  val intIn = buffer(io.int_in)
+
+  val intInMatch = intIn.bits.netId === id.U
+  io.ext_out.valid := intIn.valid && intInMatch
+  io.ext_out.bits := intIn.bits
+  io.int_out.valid := intIn.valid && !intInMatch
+  io.int_out.bits := intIn.bits
+  intIn.ready := Mux(intInMatch, io.ext_out.ready, io.int_out.ready)
+}
+
+class NetworkRing[T <: Data](
+    nIn: Int, nOut: Int, payloadTyp: T,
+    buffer: BufferParams = BufferParams.default,
+    inputFirst: Boolean = true,
+    inputMap: Option[Seq[Int]] = None,
+    outputMap: Option[Seq[Int]] = None)
+    extends NetworkInterconnect[T] {
+  val nNodes = nIn + nOut
+  val io = IO(new NetworkIO(nIn, nOut, payloadTyp, Some(nNodes)))
+
+  def mappingOK(mapping: Seq[Int], n: Int): Boolean = {
+    mapping.size == n &&
+    mapping.sorted == Seq.tabulate(n)(i => i)
+  }
+
+  require(inputMap.isEmpty  || mappingOK(inputMap.get, nIn))
+  require(outputMap.isEmpty || mappingOK(outputMap.get, nOut))
```

Comment:
It would be nice to tell the user why the `require` fails. Something like:
```suggestion
  require(outputMap.isEmpty || mappingOK(outputMap.get, nOut), "outputMap must either be None or the same length as the number of outputs")
```

Reply:
@jwright6323 I pushed a commit to add a more descriptive message. Do you think it's satisfactory?

Reply:
LGTM